libraries/AP_NavEKF/EKFGSF_yaw.cpp

```diff
@@ -0,0 +1,678 @@
+/*
+  Definition of functions used to provide a backup estimate of yaw angle
+  that doesn't use a magnetometer. Uses a bank of 3-state EKF's organised
+  as a Guassian sum filter where states are velocity N,E (m/s) and yaw angle (rad)
+
+  Written by Paul Riseborough <p_riseborough@live.com.au>
+
+  This program is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+
+#include "AP_NavEKF/EKFGSF_yaw.h"
+#include <AP_AHRS/AP_AHRS.h>
+#include <AP_Vehicle/AP_Vehicle.h>
+#include <GCS_MAVLink/GCS.h>
+
+EKFGSF_yaw::EKFGSF_yaw()
+{
+	// this flag must be false when we start
+	ahrs_tilt_aligned = false;
+
+	// these objects are initialised in initialise() before being used internally, but can be reported for logging before then 
+	memset(&GSF, 0, sizeof(GSF));
+	memset(&EKF, 0, sizeof(EKF));
+};
+
+void EKFGSF_yaw::update(const Vector3f delAng,
+                        const Vector3f delVel,
+                        const float delAngDT,
+                        const float delVelDT,
+                        bool runEKF,
+                     	float TAS)
+{
+
+	// copy to class variables
+	delta_angle = delAng;
+	delta_velocity = delVel;
+	angle_dt = delAngDT;
+	velocity_dt = delVelDT;
+	run_ekf_gsf = runEKF;
+	true_airspeed = TAS;
+
+	// Iniitialise states and only when acceleration is close to 1g to rpevent vehicle movement casuing a large initial tilt error
+	ahrs_accel = delta_velocity / fmaxf(velocity_dt, 0.001f);
```

Comment:
I've never seen us use fmaxf before.  normally we just use MAX.  This will limit the maximum update rate to 1000hz.  I guess that's OK but just want to point out that it will probably act strangely if we ever try to increase the update rate beyond that.  "strangely" in that I guess the accelerations used will simply be inaccarate.

Reply:
slight typo in "Initialise" (two "i"s).

Reply:
Typo fixed.

Reply:
I could use the average EKF rate /4  to allow for jitter

Reply:
I've pushed a change to make the limitation more obvious to integrators